Amplify/Categories/DataStore/Model/PropertyPath.swift

```diff
@@ -0,0 +1,126 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+/// Runtime information about a property. Its `name` and `parent` property reference,
+/// as well as whether the property represents a collection of the type or not.
+public struct PropertyPathMetadata {
+    public let name: String
+    public let isCollection: Bool
+    public let parent: PropertyPath?
+}
+
+/// Represents a property of a `Model`. PropertyPath is a way of representing the
+/// structure of a model with static typing, so developers can reference model
+/// properties in queries and other functionality that require them.
+public protocol PropertyPath {
+
+    /// Access the property metadata.
+    ///
+    /// **Implementation note:** this function is in place over an implicit accessor over
+    /// a property named `metadata` in order to avoid name conflict with the actual property
+    /// names that will get generate from the `Model`.
+    ///
+    /// - Returns the property metadata, that contains the name and a reference to its parent.
+    func getMetadata() -> PropertyPathMetadata
```

Comment:
How can a type in another module conform to `PropertyPath` if `PropertyPathMetadata` can't be instantiated from that module?

Reply:
It looks like these are only being used / conformed to within this module. Can `PropertyPath` and `PropertyContainerPath` be internal?

Reply:
> Can PropertyPath and PropertyContainerPath be internal?

I attempted to remove `public` up the chain of dependencies and eventually cascades up to `Model` which has to stay public
```
/// All persistent models should conform to the Model protocol.
public protocol Model: Codable {

    /// The reference to the root path. It might be `nil` if models do not support
    /// property references.
    static var rootPath: PropertyContainerPath? { get }

```